lib/bap_types/bap_const_propagation.ml

```diff
@@ -0,0 +1,93 @@
+open Core_kernel.Std
+open Bap_common
+open Bap_bil
+open Bap_visitor
+open Stmt
+
+module Var = Bap_var
+module Word = Bitvector
+module Exp_helpers = Bap_helpers.Exp
+
+let is_assigned = Bap_helpers.is_assigned
+
+(** [remove_forward_defs bil defs]
+    for every [x := exp] in [bil], where [exp] depends on some
+    [y1 .. yN] redefined later then [x],
+    removes both [x] and [y1 .. yN] from defs *)
+let remove_forward_defs bil defs =
+  (object
+    inherit [exp Var.Map.t] bil_visitor
+    method! enter_move var e defs =
+      let redefined = Set.filter (Exp_helpers.free_vars e)
+          ~f:(fun v -> is_assigned v succs) in
+      if Set.is_empty redefined then defs
+      else
+        Set.fold ~init:(Map.remove defs var) redefined
+          ~f:(fun defs v ->
+              if is_assigned v succs then Map.remove defs v
```

Comment:
`redefined`  is defined above as a set of variables s.t. that for each `v` in this set `is_assigned v succs` is true. Thus this expression will be always true. So, what you're trying to say is that `Set.fold ~init:(Map.remove defs var) redefined ~f:Map.remove`. On in other words, it is `defs / v U redefined`.

I can understand this as a claim, that a statement `v := x` kills `c U {is_assigned x : FV(x)}` from the set of constants? 
This doesn't make any sense at all. This property has nothing to do with the FV property.